lang/csharp/src/apache/main/Schema/EnumSchema.cs

```diff
@@ -103,15 +133,46 @@ internal static EnumSchema NewInstance(JToken jtok, PropertyMap props, SchemaNam
                             string doc, string defaultSymbol)
                             : base(Type.Enumeration, name, aliases, props, names, doc)
         {
-            if (null == name.Name) throw new SchemaParseException("name cannot be null for enum schema.");
+            if (null == name.Name) throw new AvroException("name cannot be null for enum schema.");
             this.Symbols = symbols;
             this.symbolMap = symbolMap;
 
             if (null != defaultSymbol && !symbolMap.ContainsKey(defaultSymbol))
-                throw new SchemaParseException($"Default symbol: {defaultSymbol} not found in symbols");
+                throw new AvroException($"Default symbol: {defaultSymbol} not found in symbols");
             Default = defaultSymbol;
         }
 
+        /// <summary>
+        /// Creates symbols map from specified list of symbols.
+        /// Symbol map contains the names of the symbols and their index.
+        /// </summary>
+        /// <param name="symbols">List of symbols</param>
+        /// <returns>Symbol map</returns>
+        /// <exception cref="AvroException">Is thrown if the symbols list contains invalid symbol name or duplicate symbols</exception>
+        private static IDictionary<string, int> CreateSymbolsMap(IEnumerable<string> symbols)
+        {
+            IDictionary<string, int> symbolMap = new Dictionary<string, int>();
+            int i = 0;
+            foreach (var symbol in symbols)
+            {
+                if (ValidateSymbolName(symbol))
+                {
+                    throw new AvroException($"Invalid symbol name: {symbol}");
+                }
+
+                if (symbolMap.ContainsKey(symbol))
+                {
+                    throw new AvroException($"Duplicate symbol: {symbol}");
+                }
+
+                symbolMap[symbol] = i++;
+            }
+
+            return symbolMap;
+        }
+
+        private static bool ValidateSymbolName(string symbol) => string.IsNullOrEmpty(symbol) || !Regex.IsMatch(symbol, "^([A-Za-z_][A-Za-z0-9_]*)$");
```

Comment:
I always have perfromance concerns when we use Regex. I know that JsonHelper.GetRequiredString(), however is that a concern here? What I am thinking is that when someone is using a schemaBuilder in runtime, this call might be called many times? I might be just too paranoid here. In a unit test I have no concern, however in the main library, it might cause unnecessary delay for a simple check. Of course if the name might be more complicated in the future, regex is probably the easiest/best solution.

IMO ValidateSymbol should throw the exception if not valid, something like this:

```
private static void ValidateSymbolName(string symbol)
{
    if (string.IsNullOrEmpty(symbol) || !Regex.IsMatch(symbol, "^([A-Za-z_][A-Za-z0-9_]*)$"))
    {
        throw new AvroException($"Invalid symbol name: {symbol}");
    }
}
```

Reply:
Validate that returns bool. Don't know what I was thinking. Fixed.

Yeah, Regex is not really good at performance (I don't like the readability either), but since it's a new feature (the parsing code doesn't use this validation) so we are not degrading performance, and since my guess that usually schemas are not created in a tight loop (usually created once and reused), it should be ok. It can be fixed if it does prove to be a problem.